cpp/farm_ng/core/prototools/event_log_writer_test.cpp

```diff
@@ -44,15 +44,140 @@ TEST(event_log, roundtrip) {
     EventLogReader reader(file);
     for (int i = 0; i < 10; ++i) {
       std::string payload;
-      auto [event, pos] = reader.readNextEvent(&payload);
+      EventLogPos pos = reader.readNextEvent(&payload);
       core::proto::Timestamp x;
       EXPECT_EQ(true, x.ParseFromString(payload));
       EXPECT_EQ(i, x.stamp());
-      EXPECT_EQ("protobuf", event.uri().scheme());
-      EXPECT_EQ("my_stamps", event.uri().path());
-      EXPECT_EQ(getHostName(), event.uri().authority());
-      EXPECT_EQ("type=farm_ng.core.proto.Timestamp", event.uri().query());
+      EXPECT_EQ("protobuf", pos.event().uri().scheme());
+      EXPECT_EQ("my_stamps", pos.event().uri().path());
+      EXPECT_EQ(getHostName(), pos.event().uri().authority());
+      EXPECT_EQ("type=farm_ng.core.proto.Timestamp", pos.event().uri().query());
     }
     EXPECT_THROW(reader.readNextEvent(), std::runtime_error);
   }
 }
+
+TEST(event_log, index) {
+  auto maybe_log_dir = createUniqueTemporaryDirectory();
+  auto log_dir = FARM_UNWRAP(maybe_log_dir);
+  auto file = log_dir / "events.log";
+  {
+    EventLogWriter writer(file);
+    core::proto::Timestamp x;
+    for (int i = 0; i < 10; ++i) {
+      x.set_stamp(i);
+      writer.write("my_stamps", x);
+    }
+  }
+  {
+    EventLogReader reader(file);
+    int i = 0;
+    for (EventLogPos const& pos : reader.getIndex()) {
+      core::proto::Timestamp x;
+      x.ParseFromString(pos.readPayload());
+      EXPECT_EQ(i, x.stamp());
+      EXPECT_EQ("protobuf", pos.event().uri().scheme());
+      EXPECT_EQ("my_stamps", pos.event().uri().path());
+      EXPECT_EQ(getHostName(), pos.event().uri().authority());
+      EXPECT_EQ("type=farm_ng.core.proto.Timestamp", pos.event().uri().query());
+      ++i;
+    }
+  }
+}
+
+TEST(event_log, time_ordered_zip) {
+  auto maybe_log_dir = createUniqueTemporaryDirectory();
+  auto log_dir = FARM_UNWRAP(maybe_log_dir);
+  auto file1 = log_dir / "events1.log";
+  auto file2 = log_dir / "events2.log";
+
+  {
+    EventLogWriter writer1(file1);
+    EventLogWriter writer2(file2);
+    core::proto::Timestamp x;
+    for (int i = 0; i < 10; ++i) {
+      x.set_stamp(i);
+      writer1.write("my_stamps1", x);
+      writer2.write("my_stamps2", x);
+    }
+  }
+  {
+    EventLogReader reader1(file1);
+    EventLogReader reader2(file2);
+    // Grab a reference write stamp from the first reader.
+    core::proto::Timestamp ref =
+        reader1.readNextEvent().event().timestamps()[0];
+    EXPECT_EQ(ref.semantics(), "log/write");
+    std::vector<EventLogPos> index = eventLogTimeOrderedIndex(
```

Comment:
This can throw, we need a catch block to document to uses canonical use of API.

(Likely there are more throwing functions below.)

Reply:
this won't throw unless the file doesn't exist (which will throw in on construction of the reader.